### PR TITLE
fix goroutine leak in watchMux

### PR DESCRIPTION
**Pull request by ikaven1024**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #4212

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
`karmada-search`: Fix lock race affects watch RestChan not close, causing client watch api to hang.
```

### CODE REVIEW

File: pkg/search/proxy/store/util.go [hunk 1]
```diff
@@ -195,9 +197,67 @@ func (w *watchMux) AddSource(watcher watch.Interface, decorator func(watch.Event
...
+		for {
+			chosen, val, ok := reflect.Select(cases)
+			if !ok {
+				return
+			}
+
+			if chosen == doneCaseIndex {
+				// Received from done chan.
+				// In fact, this will never happen
+				panic(fmt.Sprintf("unexpectedly receive from done chan: %v", val.Interface()))
```
Comment (xigang) [inline pkg/search/proxy/store/util.go:243]: If panic is triggered, does the exception need to be caught?

File: pkg/search/proxy/store/util.go [hunk 2]
```diff
@@ -195,9 +197,67 @@ func (w *watchMux) AddSource(watcher watch.Interface, decorator func(watch.Event
...
+
+	go func() {
+		defer func() {
+			for _, source := range w.sources {
+				source.watcher.Stop()
+			}
+		}()
+		defer close(w.result)
+		for {
+			chosen, val, ok := reflect.Select(cases)
```
Comment (xigang) [inline pkg/search/proxy/store/util.go:235]: The performance of `reflect.Select` is worse than that if `select`. if the scale of the federate cluster is too large, there may be performance issue.

File: pkg/search/proxy/store/util.go [hunk 3]
```diff
@@ -246,19 +257,8 @@ func (w *watchMux) startWatchSource(source watch.Interface, decorator func(watch
 		select {
 		case <-w.done:
 			return
-		default:
+		case w.result <- copyEvent:
```
Comment (xigang) [inline pkg/search/proxy/store/util.go:260]: Is there a data race when multiple goroutines write w.result channel?

File: pkg/search/proxy/store/util.go [hunk 4]
```diff
@@ -195,9 +195,21 @@ func (w *watchMux) AddSource(watcher watch.Interface, decorator func(watch.Event
...
+		source := source
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			w.startWatchSource(source.watcher, source.decorator)
+		}()
 	}
+
+	go func() {
+		// close result chan after all goroutines exit, avoiding data race.
```
Comment (XiShanYongYe-Chang) [inline pkg/search/proxy/store/util.go:209]: Just a question, why would there be data race in the previous approach?

Comment (RainbowMango) [reply pkg/search/proxy/store/util.go:209]: @ikaven1024 explained it at https://github.com/karmada-io/karmada/pull/4212#issuecomment-1805660860.

**Pull request closed without merging**